### ISSUE TITLE
takePicture preview with Nikon D7000

### ISSUE THREAD

**Issue by Okkervile** [lwille/node-gphoto2#48]
takePicture with parameter preview:true fails silently with Nikon D7000.
takePicture with parameter downloading true is working fine.

Any known issues with Nikon?

**Comment by lwille** [comment]
Any news? What about the `gp_camera_exit` feature?

**Comment by lpinca** [comment]
@lwille no news on my side, I didn't try anymore.

**Comment by gabrielstuff** [comment]
Hello there, @Iwille I've been a bit forward and now I got the image correctly draw in the canvas. @lpinca  was right the buffer was always empty because of two things.
First, buffer is only returned if downloaded, see there:  https://github.com/lwille/node-gphoto2/blob/master/src/camera.cc#L121
`} else if (capture_req->data && capture_req->download) {`

And secondly, the `capturePreview` does not generate a buffer. The buffering only happen if we call takePicture so... I had to add this snippet to the capturePreview function : 

```
if (retval == GP_OK && req->target_path.empty()) {
    retval = gp_file_get_data_and_size(file, &data, &req->length);
    if (retval == GP_OK && req->length != 0) {
      /* `gp_file_free` will call `free` on `file->data` pointer, save data */
      req->data = new char[req->length];
      memmove(const_cast<char *>(req->data), data, req->length);
    }
    data = NULL;
  }
```

Now I've almost got something... but it is a still image. I'm keeping on investigating.

thanks !

**Comment by lpinca** [comment]
@gabrielstuff it's up to the client to refresh the image by repeatedly calling `capturePreview.
As you can see here https://github.com/lwille/node-gphoto2/blob/master/examples/public/index.html#L19-L24 a new request is done every time the image is loaded.

If you are running that code and you get a still image then yes, something is wrong.

**Comment by gabrielstuff** [comment]
Yup ! I've seen that, but image won't refresh, until I use setInterval and call the function repeatidly... which is slow slow slow :)
If I click several time it update but second after seconds

**Comment by lpinca** [comment]
I don't know it could be your camera.
Is is fast enough if you use `gphoto2` directly?

**Comment by lwille** [comment]
@gabrielstuff did you try streaming the data? Should work much quicker than using a buffer.

https://github.com/lwille/node-gphoto2/blob/master/test/camera.test.coffee#L119

**Comment by gabrielstuff** [comment]
Yes, 20fps using our test software : pyying --> https://github.com/soixantecircuits/pyying

**Comment by gabrielstuff** [comment]
@lwille do you mean by streaming the buffer using pipe ? As you can see the examples won't work without the modification I just had based on the remark from @lpinca. I'm note sure to see what you mean by stream.
Currently we have two options : 

- fetching a buffer
- fetching a file with new filename every xx second

Thanks !

**Comment by lwille** [comment]
Streaming using a socket, just as in the example I linked in my previous reply.
It may look like a hack, but actually worked pretty nice - passing huge buffers from the C++ module to node seemed to have more overhead than just streaming via a socket.

Sadly, I can't help about problems with different cameras etc, as I'm not actively maintaining this module any more - the Canon EOS 550D that I used to develop is broken, and the 5D is sold ¯\_(ツ)_/¯

**Comment by jatpatel1** [comment]
Hi guys,
I was wondering if someone can help me get this running.
I'm pretty new to node.js (I come from a php, js and jquery background).
I'm basically trying to do what @gabrielstuff is doing(with my Nikon D90). 

I got all the modules installed but don't know how to start the server etc...

I did create a new js file and called `camera.takePicture()` which basically created one image and then the camera mirror locked with no errors. I had to turn it on and off. 
Do I need to do this in a loop or something to continuously update the image?

Thanks
